### ISSUE TITLE
[HGQ2 0.1.8 / torch backend] Per-step forward-path GPU memory leak (~17–44 MB/step)

### ISSUE THREAD

**Issue by EEEEEEYu** [calad0i/HGQ2#10]
## Summary

On the **torch** backend, every training step of an HGQ2 model leaks GPU memory (peak grows linearly across steps), eventually OOM-ing on long runs. The leak reproduces with **no autograd** (`torch.no_grad()` + forward-only loop), so it's in HGQ2's forward path itself, not in autograd / backward / optimizer.

The same model + scope on the **JAX** backend (`KERAS_BACKEND=jax`, `jit_compile=True`) shows **bit-identical memory across 400 steps** (allocated stays at 247 MB, peak stays at 2092 MB) — so the leak is torch-backend-specific.

A plain Keras `Conv2D + Dense` stack (no HGQ2 layers) on the torch backend is stable (+0.17 MB/step drift). So the leak is **HGQ2 + torch backend**, not Keras-on-torch in general.

## Versions

```
HGQ2: 0.1.8
keras: 3.14.1
torch: 2.8.0 + CUDA 12.8
jax: 0.10.0 (CUDA 12)        # used only for the contrast test below
tensorflow: 2.21.0 (CPU)     # used only for the contrast test below
python: 3.12
GPU: NVIDIA RTX A6000 (47.4 GiB)
```

## Reproducer

Minimal forward-only loop, no Lightning, no fancy wrapper. Two QConvT2D + two QDenseT, batch-norm and bumped bit-widths to make the forward signal non-zero (cf. the [bit-width tuning notes](#bit-width-tuning-context) below).

```python
import os
os.environ["KERAS_BACKEND"] = "torch"  # ← swap to "jax" to see the contrast

import h5py, numpy as np, torch, keras
import torch.nn as nn
from hgq.config import LayerConfigScope, QuantizerConfigScope
from hgq.layers import QConvT2D, QDenseT


def build_model():
    with (
        QuantizerConfigScope(q_type='kbi', place='weight', b0=10, i0=2,
                             overflow_mode='SAT_SYM', round_mode='RND'),
        QuantizerConfigScope(q_type='kif', place='datalane', i0=4, f0=4,
                             overflow_mode='SAT_SYM', round_mode='RND'),
        QuantizerConfigScope(q_type='kbi', place='table', b0=10, i0=2,
                             overflow_mode='SAT', homogeneous_axis=(0,),
                             round_mode='RND'),
        LayerConfigScope(enable_ebops=False, beta0=0),
    ):
        m = keras.Sequential([
            keras.layers.MaxPooling2D(2),
            QConvT2D(32, 3, n_hl=1, d_hl=4, padding='same',
                     batch_norm=True, bn_epsilon=1e-9),
            keras.layers.MaxPooling2D(2),
            QConvT2D(64, 3, n_hl=1, d_hl=4, padding='same',
                     batch_norm=True, bn_epsilon=1e-9),
            keras.layers.GlobalAveragePooling2D(),
            QDenseT(64, n_hl=1, d_hl=4, activation='relu',
                    batch_norm=True, bn_epsilon=1e-9),
            QDenseT(10, n_hl=1, d_hl=4, batch_norm=True, bn_epsilon=1e-9),
        ])
    m.build((None, 32, 32, 16))
    return m


class Wrap(nn.Module):
    def __init__(self, m):
        super().__init__()
        self.model = m
    def forward(self, x):
        return self.model(x, training=self.training)


device = "cuda"
model = Wrap(build_model()).to(device)

# Random input matching what our event-camera pipeline feeds.
x = torch.randn(32, 32, 32, 16, device=device)

torch.cuda.synchronize()
init_mb = torch.cuda.memory_allocated() / 1024 / 1024
print(f"before loop: allocated = {init_mb:.1f} MB")

model.train()
with torch.no_grad():
    for step in range(200):
        _ = model(x)
        if (step + 1) % 50 == 0:
            torch.cuda.synchronize()
            alloc = torch.cuda.memory_allocated() / 1024 / 1024
            peak = torch.cuda.max_memory_allocated() / 1024 / 1024
            print(f"step={step+1:4d}  allocated={alloc:7.1f} MB  "
                  f"+{alloc-init_mb:+6.1f} MB  peak={peak:7.1f} MB")
            torch.cuda.reset_peak_memory_stats()
```

### `KERAS_BACKEND=torch`, forward-only, `no_grad`

```
before loop: allocated =   170.0 MB
step=  50  allocated=  4855.7 MB  +4685.7 MB  peak=  6490.6 MB
step= 100  allocated=  5701.5 MB  +5531.6 MB  peak=  8959.6 MB
step= 150  allocated=  9617.7 MB  +9447.7 MB  peak= 12344.1 MB
step= 200  allocated= 11506.8 MB +11336.9 MB  peak= 14620.9 MB
```

`+11.3 GB` over 150 steps, linear growth. Continuing the run OOMs an RTX A6000 (47 GB) around step ~600.

### Same code, `KERAS_BACKEND=jax`, `model.compile(jit_compile=True)`

(Forward path is wrapped in a `jax.jit` via `model.compile` / `model.train_on_batch`.)

```
before loop: allocated=170.0 MB
step= 50  allocated=246.9 MB  +76.9 MB  peak=2091.8 MB  loss=2.3950
step=100  allocated=246.9 MB  +76.9 MB  peak=2091.8 MB  loss=2.0844
step=150  allocated=246.9 MB  +76.9 MB  peak=2091.8 MB  loss=1.8274
step=200  allocated=246.9 MB  +76.9 MB  peak=2091.8 MB  loss=1.6227
step=400  allocated=246.9 MB  +76.9 MB  peak=2091.8 MB  loss=1.1561
```

Bit-identical memory across 400 steps. Loss decreases.

### Plain `keras.Conv2D` + `keras.Dense` (no HGQ2), `KERAS_BACKEND=torch`

```
before loop: allocated=128.1 MB
step=  50  allocated=148.8 MB  +20.7 MB  peak=154.3 MB
step= 100  allocated=158.8 MB  +30.7 MB  peak=162.3 MB
step= 150  allocated=163.8 MB  +35.7 MB  peak=171.3 MB
step= 200  allocated=173.8 MB  +45.7 MB  peak=179.3 MB
```

~0.17 MB/step drift — essentially stable. So Keras-on-torch by itself is fine; the leak shows up only with HGQ2 layers in the graph.

## Variables that don't change the leak (on torch backend)

- `enable_ebops=False` → slight reduction (27 → 17 MB/step) but leak persists.
- `batch_norm=False` → no improvement (sometimes worse, ~44 MB/step).
- `overflow_mode='WRAP'` vs `'SAT_SYM'` → no change.
- `torch.cuda.empty_cache()` between steps → no reclaim. The memory is genuinely held by live Python references, not just fragmentation.
- Switching backend to `KERAS_BACKEND=tensorflow` (CPU only on our box) reproduces the same kind of zero-output bug we saw before the bit-width fixes, but we couldn't measure GPU memory there because TF didn't pick up GPUs. The diagnostic point is that **JAX is leak-free**, so the issue is specific to the **torch backend's variable assign / runtime**, not to HGQ2's algorithm.

## Hypothesis

HGQ2's per-call state mutations on per-element bit-width Variables (`_i.assign(...)`, `_f.assign(...)`) and on BatchNormalization's `moving_mean.assign(...)` / `moving_variance.assign(...)` — these go through Keras's `Variable.assign` API, which under the torch backend writes a new tensor each call instead of mutating in place. Old tensor references aren't released immediately by Python's GC, so they accumulate at roughly the rate of the bit-width state tensor size per layer per step. The pattern matches the observed magnitude (~5 MB per HGQ2 layer per step at our shapes).

The fact that the leak rate scales with the bit-width state-tensor size (per-element granularity → more leak) and that JAX's XLA-traced path doesn't leak (because the compiled graph plans the buffers once and reuses them) is consistent with this.

## Bit-width tuning context

Worth noting (it took us a while to find): the README MNIST example uses `q_type='kif'` for `place='weight'`, but `place='weight'` defaults to `q_type='kbi'`, so the scope is a silent no-op. The actual kbi defaults `b0=4, i0=2` give 2 fractional bits → resolution 0.25, which rounds Glorot-init kernels of any non-tiny conv to zero on the first forward (we hit this with a `[3, 3, 16, 32]` conv → silent zero output, val_loss frozen at `ln(num_classes)`). The reproducer above uses `b0=10, i0=2` (8 fractional bits) which fixes the zero-output issue separately from the leak.

We can split this into a second issue if useful — happy to do so.

## What we did as a workaround

Migrated the whole pipeline to `KERAS_BACKEND=jax` + `model.compile(jit_compile=True)`. Confirmed full-epoch training is stable (val top-1 0.961 on a TBR-encoded N-MNIST baseline in 67 s/epoch) at the same batch size and architecture that OOM'd on torch.

**Comment by calad0i** [comment]
Thanks for the report. `getting_started.md` doc issue is fixed.

Regarding the mem leak issue, it appears to be upstream in native keras. The shown memory leak like behavior can be mitigated with `gc.collect()`, which suggests that the intermediate tensors are reachable by nested (orphaned) python references rather than real orphaned allocated blocks.

In your provided example, in my local test case the default gc behavior kicks in at around 500 steps and the allocated memory drops suddenly around step 500. The same seesaw mem util pattern also appears in native keras w/ einsumdense. While HGQ creates large intermediate and amplifies the issue, it does not appear to be the cause of the issue. Checked locally with `keras==3.14.1` and `torch==2.12.0` on `python==3.12.13` (p.s., py314+ has [a separate GC issue](https://discuss.python.org/t/reverting-the-incremental-gc-in-python-3-14-and-3-15/107014)). 

Not directed to the reported issue, but it is not recommended to use `torch` with hgq or using a custom training loop unless is particular strong reasons doing so. In most cases, `model.fit` with jax (or tf) with jit compilation provides far superior performance, and as static dataflow graph is generally enforced on hardware anyway, there is no significant drawback doing so. Regarding the model itself. it is likely too large to be implementable. For a event camera input, the input width is likely small and using LUT-Conv does not appear to be a good choice (and the channel number is likely too high: each conv kernel can be as large as 9*32->, and multiple ).

**Comment by calad0i** [comment]
Closing as completed; please re-open if you have further questions
